Match insight

RaysDev opened the pull request:
* This is a…
    * [ ] Bug fix
    * [x] Feature addition
    * [ ] Refactoring
    * [ ] Minor / simple change (like a typo)
    * [ ] Other

**#Summary**
Minor adjustment to scoring logic, and when matching file to AcoustId/musicbrainz results, provides insight to user on the scoring/final selection process.

**# Problem**
AcoustID/Musicbrainz can return multiple possibilities in their response, and we need to match the input file to just one of them.
1. some results have no release date, which are less desirable to match to.

2. while user can leverage the sliders in the preferences screen, to lower the priority of certain "release types," they have no way to completely exclude it.  since final result is based on several factors, these undesirable types still get chosen by picard.  I found this particularly troublesome with compilations.  I want to tag my mp3 with the original album, not a recent compilation.

3. I've observed that sometimes acoustid will return over hundred results for Picard to pick from.  While picard has a great algorithm for picking one.  I found it often did not chose the one I preferred (ie the original album for the specific recording).  I was left wondering why, and what could I tweak in my mp3 file's tags to get Picard to make the right match.  There was no information available.

**# Solution**
1. higher priority should be given to results that have a non-null release date.

2. if the slider in preferences screen is set to zero (far left) for a given Release Type, then Picard should never have its final pick be of that release type.  If user just wants to de-prioritize it, they can move slider slightly right of the far left.

3.  First I added detailed logging of the matching (ie scoring) process so I could see why it was picking a particular result.  But few users would want to dig through logs, and it was not time efficient when you have a lot of tracks.  Further, having the information is nice, but ultimately if we did not like Picard's selection, then we want to be able to select an alternative.  So I moved the info from the logs into a new dialog.  Available by right clicking on the track in Right side pane, and choosing "Match Target".  While there is already a "Other Versions" options, its very limited.  This new dialog, not only shows the possible results, but also has filtering/sorting capability from the column header.  So user can narrow down the results, and then click on one to apply it to the track.

phw commented:
Thanks for the PR. Since this is a rather large change I won't merge it before the 2.3 final release.

I will give it a proper review later, but can you fix the coding style issues reported by flake8 (see e.g. https://github.com/metabrainz/picard/pull/1476/checks?check_run_id=447565225)?

zas commented:
While I can see the overall idea, the whole PR has to be reworked before we can even review it:

- you need to break down things in smaller commits, with proper comments
- read and respect https://github.com/metabrainz/picard/blob/master/CONTRIBUTING.md
- remove all unrelated changes (e.g. what's the point of this commit cc331cf31d623e491eb1015b8a2936b0aed9b359 which basically change all lines of metadata.py ??)
- rebase once you cleaned up

zas commented on picard/metadata.py:
```diff
@@ -276,11 +294,70 @@ def compare_to_track(self, track, weights):
+                    medium = release['media'][0]
+                    med_fmt = medium['format'] if 'format' in medium else ""
+                    med_pos = medium['position'] if 'position' in medium else 0
+                    med_trk_cnt = medium['track-count'] if 'track-count' in medium else 0
```
What about initializing values instead and just use `get()` ?

```python
med_trk_cnt = 0
if 'media' in release:
    medium = release['media'][0]
    med_trk_cnt = medium.get('track-count', 0)
```

zas commented on picard/metadata.py:
```diff
@@ -276,11 +294,70 @@ def compare_to_track(self, track, weights):
+                    med_pos = medium['position'] if 'position' in medium else 0
+                    med_trk_cnt = medium['track-count'] if 'track-count' in medium else 0
+                    trk_pos = medium['tracks'].get('position', '') if ('tracks' in medium) else 0
+                    trk_id = medium['tracks'].get('id', '') if ('tracks' in medium) else ""
```
Useless parenthesis

zas commented on picard/metadata.py:
```diff
@@ -276,11 +294,70 @@ def compare_to_track(self, track, weights):
+                                           rel_id, rel_ctry, rel_dt,
+                                           med_fmt, med_pos, med_tot_cnt,
+                                           trk_id, trk_pos, med_trk_cnt, rel_tot_trk_cnt,
+                                           parts, release_parts, recording_base_score)
```
I'd pass an object instead

zas commented on picard/ui/mainwindow.py:
```diff
@@ -1080,6 +1096,14 @@ def update_actions(self):
+                self.track_match_details_action.setEnabled(have_objects)
+            else:
+                self.track_match_details_action.setEnabled(False)
+        except Exception:
```
Catchall exception again

zas commented on picard/ui/mainwindow.py:
```diff
@@ -1080,6 +1096,14 @@ def update_actions(self):
+            else:
+                self.track_match_details_action.setEnabled(False)
+        except Exception:
+            self.track_match_details_action.setEnabled(False)
```
What about storing boolean in a variable and call `self.track_match_details_action.setEnabled(is_enabled)`?

zas commented on picard/ui/matchdetailsdialog/__init__.py:
```diff
@@ -0,0 +1,379 @@
+# Copyright (C) 2016 Rahul Raturi
+# Copyright (C) 2018 Laurent Monin
+# Copyright (C) 2019 Philipp Wolfer
+# Copyright (C) 2020 Ray Bouchard
```
This is a new file, so it doesn't make sense to add copyright from other authors and previous years.

zas commented on picard/ui/matchdetailsdialog/__init__.py:
```diff
@@ -0,0 +1,379 @@
+        self.horizontalHeader().setContextMenuPolicy(QtCore.Qt.CustomContextMenu)
+        self.horizontalHeader().customContextMenuRequested.connect(self.show_header_context_menu)
+        self.keywords = dict([(i, []) for i in range(self.columnCount())])
+        self.checkBoxs = []
```
checkboxes?

zas commented on picard/ui/matchdetailsdialog/__init__.py:
```diff
@@ -0,0 +1,379 @@
+                self.checkBoxs.append(checkBox)
+
+            btn = QDialogButtonBox(QDialogButtonBox.Ok | QDialogButtonBox.Cancel,
+                                         QtCore.Qt.Horizontal, self.menu)
```
Indentation

zas commented on picard/ui/matchdetailsdialog/__init__.py:
```diff
@@ -0,0 +1,379 @@
+            posY = headerPos.y() + self.horizontalHeader().height()
+            posX = headerPos.x() + self.horizontalHeader().sectionPosition(self.col)
+            self.menu.exec_(QtCore.QPoint(posX, posY))
+        except Exception:
```
Another catchall Exception

zas commented on picard/ui/matchdetailsdialog/__init__.py:
```diff
@@ -0,0 +1,379 @@
+                did_filter = True
+
+        # add decorator to show filtered.
+        if (did_filter):
```
Parenthesis

zas commented on picard/ui/matchdetailsdialog/__init__.py:
```diff
@@ -0,0 +1,379 @@
+            for key, value in columnsShow.items():
+                self.setRowHidden(key, not value)
+
+        except Exception:
```
Catchall exception

zas commented on picard/ui/matchdetailsdialog/__init__.py:
```diff
@@ -0,0 +1,379 @@
+        return self.__colkeys.index(colname)
+
+    def set_acoustid_lkp_url_edit(self, p_acoustid_lkp_url):
+        self.acoustid_lkp_url.setText(_("&#160;<a href='" + p_acoustid_lkp_url + "'>Do AcoustId Fingerprint Lkp</a> (Below jSON copied to clipboard, use this link to <a href='https://jsongrid.com/json-grid'>Format</a> it)"))
```
What's `lkp`?

zas commented on picard/ui/matchdetailsdialog/matchdetails.py:
```diff
@@ -0,0 +1,205 @@
+            ('acoust_id',            _("AcoustId ID")),
+            ('rec_id',               _("Recording ID")),
+            ('rec_title',            _("Recording Title")),
+            ('rec_len',              _("Rec Len")),
```
Marking such string for translation doesn't help translators.

zas commented on picard/ui/matchdetailsdialog/matchdetails.py:
```diff
@@ -0,0 +1,205 @@
+                self.set_table_item(row, 'rel_tot_trk_cnt',      entry, "rel_tot_trk_cnt")
+                self.set_table_item(row, 'parts',                entry, "parts")
+                self.set_table_item(row, 'release_parts',        entry, "release_parts")
+                self.set_table_item(row, 'recording_base_score', entry, "recording_base_score")
```
Looping for columns?

Something like:

```python
columns = (''nbr', 'pfx_str', ....)
for name in columns:
   self.set_table_item(row, name, entry, name) 
```

zas commented on picard/ui/matchdetailsdialog/matchdetails.py:
```diff
@@ -0,0 +1,205 @@
+                   "%s\t%s\t%s\t"
+                   "%s\t%2d\t%2d\t"
+                   "%s\t%2d\t%2d\t%2d\t"
+                   "%s\t%s\t%5.3f")
```
https://www.python.org/dev/peps/pep-3101/

zas commented on picard/ui/matchdetailsdialog/matchdetails.py:
```diff
@@ -0,0 +1,205 @@
+                   "%s\t%2d\t%2d\t%2d\t"
+                   "%s\t%s\t%5.3f")
+
+        with open("picard_match_details.txt", "w") as text_file:
```
Where does it write?

zas commented on picard/util/__init__.py:
```diff
@@ -512,7 +515,14 @@ def load_json(data):
 
 def parse_json(reply):
-    return load_json(reply.readAll())
+    global last_json_response
```
What is the goal of this?

RaysDev commented:
Thanks for considering my pull request.  I'm working to incorporate your feedback.  I've done a lot of it today and pushed it to github.  When I'm done, do you want a new pull request, or continue working off of this one?

Its my first time on this and I did not have isort and flake8 running locally.  So I did no see all the formatting errors til after I submitted Pull Request.  I now have those tools working locally, and I was able to clean it up per your standards.  I also squashed all the commits I had done for formatting into one.

Sorry about the "noemail."  Locally, when I check my git history it was showing my actual name with noemail@gmail.com similar as I see for other committers, but for some reason, github only showed the "noemail".  I've updated it, and latest commits show my github profile.

I originally had 2 commits  1 for the logic change on the scoring, and second for the new dialog box.  I can split the logging from the dialog box into separate commits, but unless I'm unaware of some tricks, its quite a bit of work.  so I'm just asking for confirmation that is what you want before I start over with new repo?

I had read Contributing.md, I tried to honor it.  If there is anything I should change beyond feedback you have already given, then please let me know.

For the specific feedback you provided on the code.  I'm listing responses all here.  If you prefer/require I update each comment  you provided in github, I can do that too.  Please just let me know.


For commit, cc331cf, that you asked about.  somehow \r\n style newlines got introduced to that one file.  Although I do a file compare before committing, it was ignoring whitespace differences, so I did not see it.  I have fixed it.


For file.py,  line 24,  I added back in the "import os" per your requst.  I had removed it because IDE said it was not used.

I'll work on converting match_details_log to an object, give me little more time for this one.


I had match_details log, logging at info level, because it seemed key information to me, that I always wanted to see.  If it only shows at debug level, then its hard to find among all the other logging.  I understand that may not be the case for everyone.  We can change it if you prefer.

metadata.py: I had catchall exception, because match logging is naturally not core function of the program, and I did not want to cause any crashes if just logging failed.  I could put a few anticipaded exception types there, but if something else went wrong later, it would still crash. 

removed 3x inline 'date' element check per your request

converted inline ifs, to get() where you suggested

removed extra parenthesis where you suggested

converted to boolean as suggested:    self.track_match_details_action.setEnabled(is_enabled)

had other names/years in copyright of the new matchdetailsdialog/__init__.py since I based this file on the searchdialog, I was just giving credit and showing the history.  I removed per your suggestion

in matchdetailsdialog/__init__.py   I renamed checkboxes to filter_checkboxes per your request to be more descriptive

in matchdetailsdialog/__init__.py   I unabbreviated lkp=lookup per your suggestion

match_details.py removed the _() as suggested

util/__init__.py :  parse_json(reply) question regarding purpose of my change.  this was the only place I could find to intercept the unmodified response from the acoustId service, so that I could log it.  this method both consumes the data off the wire, and returns a modified/parsed version of it.  so rest of program has no way of seeing actual json response.  My change basically captures the response json, and makes it avaiable to rest of program.

zas commented:
@RaysDev continue on this PR, you can always rebase, etc... overall I like what you do here (exposing obscure internals of Picard matching), this is quite needed.
Once the code will be cleaner, we'll see if this needs to be splitted in more PRs or whatever, but for now just focus on being more familiar with Picard's coding style, etc.. so usual suspects (mainly @phw and me) can do a full review and testing.

That's a lot of changes at once, so there will be quite a lot rounds before any merge, that's not unusual ;)

Thanks for contributing.

zas commented:
> Thanks for considering my pull request. I'm working to incorporate your feedback. I've done a lot of it today and pushed it to github. When I'm done, do you want a new pull request, or continue working off of this one?
> 
> Its my first time on this and I did not have isort and flake8 running locally. So I did no see all the formatting errors til after I submitted Pull Request. I now have those tools working locally, and I was able to clean it up per your standards. I also squashed all the commits I had done for formatting into one.
> 
> Sorry about the "noemail." Locally, when I check my git history it was showing my actual name with [noemail@gmail.com](mailto:noemail@gmail.com) similar as I see for other committers, but for some reason, github only showed the "noemail". I've updated it, and latest commits show my github profile.
> 
> I originally had 2 commits 1 for the logic change on the scoring, and second for the new dialog box. I can split the logging from the dialog box into separate commits, but unless I'm unaware of some tricks, its quite a bit of work. so I'm just asking for confirmation that is what you want before I start over with new repo?
> 

Yes, please split this in smaller (logical) bits.
Sorry for the work it requires, but if you want a chance to have your changes merged that's something very required. One PR can contain different linked changes, but we usually appreciate smaller incremental patches.


> I had read Contributing.md, I tried to honor it. If there is anything I should change beyond feedback you have already given, then please let me know.
> 

Check Codacy errors too.

> For the specific feedback you provided on the code. I'm listing responses all here. If you prefer/require I update each comment you provided in github, I can do that too. Please just let me know.
> 

Usually it is much better to address each comment individually. It fits better in the github workflow.

> For commit, [cc331cf](https://github.com/metabrainz/picard/commit/cc331cf31d623e491eb1015b8a2936b0aed9b359), that you asked about. somehow \r\n style newlines got introduced to that one file. Although I do a file compare before committing, it was ignoring whitespace differences, so I did not see it. I have fixed it.
> 

Yup, this kind of commit (changing all lines for nothing) are very bad, and have to be totally removed from git history.

> For file.py, line 24, I added back in the "import os" per your requst. I had removed it because IDE said it was not used.
>

You can always submit a separate PR for unrelated changes you want to make.
 
> I'll work on converting match_details_log to an object, give me little more time for this one.
> 
> I had match_details log, logging at info level, because it seemed key information to me, that I always wanted to see. If it only shows at debug level, then its hard to find among all the other logging. I understand that may not be the case for everyone. We can change it if you prefer.
> 

Info level is meant to be seen by normal users, so for this debug level is much better, but you can use a common prefix for all your log lines so it is easier to grep.

> metadata.py: I had catchall exception, because match logging is naturally not core function of the program, and I did not want to cause any crashes if just logging failed. I could put a few anticipaded exception types there, but if something else went wrong later, it would still crash.

Catchall exception should be reserved to very specific cases, not to prevent potential random crashes.

> 
> removed 3x inline 'date' element check per your request
> 
> converted inline ifs, to get() where you suggested
> 
> removed extra parenthesis where you suggested
> 
> converted to boolean as suggested: self.track_match_details_action.setEnabled(is_enabled)
> 
> had other names/years in copyright of the new matchdetailsdialog/**init**.py since I based this file on the searchdialog, I was just giving credit and showing the history. I removed per your suggestion
> 
> in matchdetailsdialog/**init**.py I renamed checkboxes to filter_checkboxes per your request to be more descriptive
> 
> in matchdetailsdialog/**init**.py I unabbreviated lkp=lookup per your suggestion
> 
> match_details.py removed the _() as suggested

On this point, what I meant is that abbreviated text isn't practical for translators, and, as such, should be avoided. We'll see what to do with that, once we'll be able to test the PR.


> 
> util/**init**.py : parse_json(reply) question regarding purpose of my change. this was the only place I could find to intercept the unmodified response from the acoustId service, so that I could log it. this method both consumes the data off the wire, and returns a modified/parsed version of it. so rest of program has no way of seeing actual json response. My change basically captures the response json, and makes it avaiable to rest of program.

Yes, I see the point, what is bothering me is the use of `global` to do that, you can keep the code as is until we think about a better solution.

RaysDev commented:
Hello,  I believe I have address all of your feedback, and its passing all the automated checks now.  Only the "Exception catch alls" remain, as I'm no sure yet, which narrower exceptions to replace it with.  If you could please take another review.  Note, I still have it as 2 commits, one for the scoring change, and one for the "Match Details" dialog.  I'll take a look tomorrow at options for breaking it up smaller commits.

zas commented:
> Hello, I believe I have address all of your feedback, and its passing all the automated checks now. Only the "Exception catch alls" remain, as I'm no sure yet, which narrower exceptions to replace it with. If you could please take another review. Note, I still have it as 2 commits, one for the scoring change, and one for the "Match Details" dialog. I'll take a look tomorrow at options for breaking it up smaller commits.

Thanks.
You still have changes to `metadata.py` and `resources.py` to revert (prolly newlines).
Also what's the source of this target image you added? license?

RaysDev commented:
I'm working on breaking into smaller pieces now.  **One question/clarification I need.  you are looking for 3 new PRs right?  not on PR with 3 commits.**

phw commented:
> you are looking for 3 new PRs right? 

Yes, exactly. If 2e break this down into functional separate PRs we can review and merge them independently. The logging change might conflict in a minor way with some of the other PRs, but that can be handled.

zas commented:
I see various issues with this PR (apart the overall style):
- the dialog is great for people wanting to debug matching, but useless to most normal users
- logging to files shouldn't be on by default, and path should be configurable: what about "save infos as tsv file" button or the like, with a dialog to choose path?
- access to json should be possible, yes, but I'm against linking to a third-party website to format it, especially since one can just load the document in a modern browser to display it for human reading: what about "save json to file" and "open json in browser" or the like?
- overall, most of the code is related to debugging, though it also changes behaviours, a clear separation is needed (different PRs)
- using globals is definitively not the way to go, better and thread-safe ways have to be found

Added features are imho very useful, though this needs more work.

mineo commented on picard/webservice/__init__.py:
```diff
@@ -572,6 +577,22 @@ def get_response_parser(cls, response_type):
+        WARNING: This is not thread-safe, so if multiple threads are making Webservice calls concurrently, they could get the wrong results here.
+        But given we have the rate-limiter, we should be fine until this is refactored.
+        """
+        return last_raw_reply
```
This only works by accident - `WebService._handle_reply` assigns to the _global_ variable `last_raw_reply`, not the one member of the `WebService` class. This should be `return WebService.last_raw_reply` (likewise for the assignments to it).

RaysDev replied:
The Match dialog is intended to show results of matching process.  So in middle pane, you must first either "Scan" which uses AcoustId, or "Lookup" which uses MusicBrainz.  Then the right click menu option is available on right pane to show all the results of that match, and allows you to choose a different result if desired.

Good catch on the last_raw_reply, I've fixed it.

RaysDev commented:
> You still have changes to `metadata.py` and `resources.py` to revert (prolly newlines).
> Also what's the source of this target image you added? license?

metadata.py  - I had fixed the newlines in second commit.  It is also fixed now in the 3 new smaller PRs.

resources.py I did not manually modify this.  I believe its a binary (ok hex) generated file. which is why it shows the whole file changed.  (I'm guessing its because of the new icons added for the right-click menu option).

The icon comes from here, which says "basic license" and is by "Free Pikk"  so I think we are ok, but check it out and if not we could choose a different icon.
http://www.myiconfinder.com/icon/target-archery-aim-archer-bullseye-game-sport/19204#.48

RaysDev commented:
Per above comments...

I expanded the cryptic/abbreviated column headers.  Really I was trying to keep from using wide column header on a 1-2 digit column value.  It was making table too wide, but I got word wrapping in the headers working, so its fine now even without the abbreviations.

I fixed the concatenation TypeError phw reported.  

Export file: no longer automatic, instead push button and use file save dialog to pick location.

Removed link to format json.  While I understand you don't want Picard endorsing any particular sites,  the formatting they do is table/grid based, and much easier to view than what browsers present.

Moved logic for last_json from util to WebService.  it’s a more logical place to track the last raw http response than a utility class.  Agreed that the global nature is NOT a good solution, but after studying the code for long time, I did not see any way to do this WITHOUT a lot of refactoring. imho, WebService should focus on http traffic and not manipulate the response before application has chance to see it.  I considered adding it to the handler() but felt that might be too intrusive at this point in time.  so I documented it and labeled it for future optimization.

RaysDev commented:
The new Match dialog is not intended for debugging.  Rather its intended to be a way for normal users to pick an alternative track from the many results sent back by AcoustId ("Scan") and MusicBrainz ("Lookup").    Hence the sorting and filtering capabilities in the table/grid, and ability to apply selection to current item.

Kind of like the "Similar Tracks" feature, but you don't have to type in a query
Kind of like the "Other Versions" but it handles when match was to wrong song or releasegroup.

From my experience, I have an unidentified file.  AcoustID search may make my preferred match, but if it only gets in the ballpark, then this dialog allows me to zero in on exact match.  Particularly useful when a "recording" is on multiple "releases."

So while my efforts started out debugging why Picard did not pick the right song, it turned into an actionable tool to accurately complete the tagging of my files.

RaysDev commented:
> 1. The logging and match details changes (the majority of this PR)
> 2. Changes to converting the AcoustID metadata
> 3. The change to completely ignore release types set to zero scoring

**I have submitted the 3 separate PRs,  Split up as suggested by phw**
